src/renderer/generic.zig

```diff
@@ -0,0 +1,2873 @@
+                }
+            }
+        }
+
+        /// Callback when the window is visible or occluded.
+        ///
+        /// Must be called on the render thread.
+        pub fn setVisible(self: *Self, visible: bool) void {
+            // If we're not visible, then we want to stop the display link
+            // because it is a waste of resources and we can move to pure
+            // change-driven updates.
+            if (comptime DisplayLink != void) link: {
+                const display_link = self.display_link orelse break :link;
+                if (visible and self.focused) {
+                    display_link.start() catch {};
+                } else {
+                    display_link.stop() catch {};
+                }
+            }
+        }
+
+        /// Set the new font grid.
+        ///
+        /// Must be called on the render thread.
+        pub fn setFontGrid(self: *Self, grid: *font.SharedGrid) void {
+            self.draw_mutex.lock();
+            defer self.draw_mutex.unlock();
+
+            // Update our grid
+            self.font_grid = grid;
+
+            // Update all our textures so that they sync on the next frame.
+            // We can modify this without a lock because the GPU does not
+            // touch this data.
+            for (&self.swap_chain.frames) |*frame| {
+                frame.grayscale_modified = 0;
+                frame.color_modified = 0;
+            }
+
+            // Get our metrics from the grid. This doesn't require a lock because
+            // the metrics are never recalculated.
+            const metrics = grid.metrics;
+            self.grid_metrics = metrics;
+
+            // Reset our shaper cache. If our font changed (not just the size) then
+            // the data in the shaper cache may be invalid and cannot be used, so we
+            // always clear the cache just in case.
+            const font_shaper_cache = font.ShaperCache.init();
+            self.font_shaper_cache.deinit(self.alloc);
+            self.font_shaper_cache = font_shaper_cache;
+
+            // Update cell size.
+            self.size.cell = .{
+                .width = metrics.cell_width,
+                .height = metrics.cell_height,
+            };
+
+            // Update relevant uniforms
+            self.updateFontGridUniforms();
+        }
+
+        /// Update uniforms that are based on the font grid.
+        ///
+        /// Caller must hold the draw mutex.
+        fn updateFontGridUniforms(self: *Self) void {
+            self.uniforms.cell_size = .{
+                @floatFromInt(self.grid_metrics.cell_width),
+                @floatFromInt(self.grid_metrics.cell_height),
+            };
+        }
+
+        /// Update the frame data.
+        pub fn updateFrame(
+            self: *Self,
+            surface: *apprt.Surface,
+            state: *renderer.State,
+            cursor_blink_visible: bool,
+        ) !void {
+            _ = surface;
```

Comment:
Now that this is all unified we can probably remove this paramater all the way up.